presto-orc/src/main/java/com/facebook/presto/orc/writer/SegmentedSliceBlockBuilder.java

```diff
@@ -0,0 +1,416 @@
+    }
+
+    @Override
+    public long getPositionsSizeInBytes(boolean[] positions)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    private long getClosedSlicesRetainedSize()
+    {
+        return sizeOf(closedSlices) + closedSlicesRetainedSize;
+    }
+
+    private long getOffsetsRetainedSize()
+    {
+        return sizeOf(offsets) + (openSegment + 1) * sizeOf(offsets[0]);
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        long offsetsSize = getOffsetsRetainedSize();
+        long closedSlicesSize = getClosedSlicesRetainedSize();
+        return INSTANCE_SIZE + openSliceOutput.getRetainedSize() + offsetsSize + closedSlicesSize;
+    }
+
+    @Override
+    public void retainedBytesForEachPart(BiConsumer<Object, Long> consumer)
+    {
+        consumer.accept(openSliceOutput, openSliceOutput.getRetainedSize());
+        consumer.accept(offsets, getOffsetsRetainedSize());
+        consumer.accept(closedSlices, getClosedSlicesRetainedSize());
+        consumer.accept(this, (long) INSTANCE_SIZE);
+    }
+
+    @Override
+    public Block copyPositions(int[] positions, int offset, int length)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public BlockBuilder writeByte(int value)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public BlockBuilder writeShort(int value)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public BlockBuilder writeInt(int value)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public BlockBuilder writeLong(long value)
+    {
+        throw new UnsupportedOperationException(getClass().getName());
+    }
+
+    @Override
+    public BlockBuilder writeBytes(Slice source, int sourceIndex, int length)
+    {
+        if (openSegmentOffset == 0) {
+            // Expand Segments if necessary.
+            if (openSegment >= offsets.length) {
+                int newCapacity = Math.max(openSegment + 1, (int) (offsets.length * 1.5));
+                closedSlices = Arrays.copyOf(closedSlices, newCapacity);
+                offsets = Arrays.copyOf(offsets, newCapacity);
+            }
+
+            if (offsets[openSegment] == null) {
+                offsets[openSegment] = new int[SegmentHelper.SEGMENT_SIZE + 1];
+            }
+        }
+        openSliceOutput.writeBytes(source, sourceIndex, length);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder closeEntry()
+    {
+        openSegmentOffset++;
+        offsets[openSegment][openSegmentOffset] = openSliceOutput.size();
+        if (openSegmentOffset == SegmentHelper.SEGMENT_SIZE) {
+            // Add the current finalized slice to closedSlices
+            Slice slice = openSliceOutput.copySlice();
```

Comment:
We do a copy to save space? We usually don't call this method to avoid heavy GC. Maybe `slice()` is good enough.

Reply:
Slice() gives the view of the object. After the segment is full, the bytes[] are copied to the copySllice and the dynamicSlliceOuptut is reset and reused for the new segment.